src/librustc_trans/back/link.rs

```diff
@@ -939,6 +932,12 @@ fn add_upstream_rust_crates(cmd: &mut Linker,
         // symbols from the dylib.
         let src = sess.cstore.used_crate_source(cnum);
         match data[cnum as usize - 1] {
+            // We must always link the `compiler_builtins` crate statically. Even if it was already
+            // "included" in a dylib (e.g. `libstd` when `-C prefer-dynamic` is used)
+            _ if sess.cstore.is_compiler_builtins(cnum) => {
+                add_static_crate(cmd, sess, tmpdir, crate_type,
+                                 &src.rlib.unwrap().0, sess.cstore.is_no_builtins(cnum))
```

Comment:
Oh actually I think this may be where the bug is. Not only do we always need to add the crate statically, but we need to always put it as the last crate linked.

Could the `cnum` be saved off to get passed at the end of this loop?